controllers/nstemplateset/namespaces.go

```diff
@@ -275,3 +287,97 @@ func getNamespaceName(request reconcile.Request) (string, error) {
 	}
 	return namespace, nil
 }
+
+// isUpToDateAndProvisioned checks if the obj has the correct Template Reference Label.
+// If so, it processes the tier template to get the expected roles and rolebindings and then checks if they are actually present in the namespace.
+func (r *namespacesManager) isUpToDateAndProvisioned(ns *corev1.Namespace, tierTemplate *tierTemplate) (bool, error) {
+	if ns.GetLabels() != nil &&
+		ns.GetLabels()[toolchainv1alpha1.TierLabelKey] == tierTemplate.tierName &&
+		ns.GetLabels()[toolchainv1alpha1.TemplateRefLabelKey] == tierTemplate.templateRef {
+
+		newObjs, err := tierTemplate.process(r.Scheme, ns.GetLabels()[toolchainv1alpha1.OwnerLabelKey], template.RetainAllButNamespaces)
+		if err != nil {
+			return false, err
+		}
+		processedRoles := []runtimeclient.Object{}
+		processedRoleBindings := []runtimeclient.Object{}
+		roleList := rbac.RoleList{}
+		rolebindingList := rbac.RoleBindingList{}
+		if err = r.AllNamespacesClient.List(context.TODO(), &roleList, runtimeclient.InNamespace(ns.GetName())); err != nil {
+			return false, err
+		}
+
+		if err = r.AllNamespacesClient.List(context.TODO(), &rolebindingList, runtimeclient.InNamespace(ns.GetName())); err != nil {
+			return false, err
+		}
+
+		for _, obj := range newObjs {
+			switch obj.GetObjectKind().GroupVersionKind().Kind {
+			case "Role":
+				processedRoles = append(processedRoles, obj)
+			case "RoleBinding":
+				processedRoleBindings = append(processedRoleBindings, obj)
+			}
+		}
+		// get the owner name from namespace
+		owner, exists := ns.GetLabels()[toolchainv1alpha1.OwnerLabelKey]
+		if !exists {
+			return false, fmt.Errorf("namespace doesn't have owner label")
+		}
+		//Check the names of the roles and roleBindings as well
+		for _, role := range processedRoles {
+			if found, err := r.containsRole(roleList.Items, role, owner); !found {
+				return false, err
+			}
+		}
+
+		for _, rolebinding := range processedRoleBindings {
+			if found, err := r.containsRoleBindings(rolebindingList.Items, rolebinding, owner); !found {
+				return false, err
+			}
+		}
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (r *namespacesManager) containsRole(list []rbac.Role, obj runtimeclient.Object, owner string) (bool, error) {
+	for _, val := range list {
+		if obj.GetObjectKind().GroupVersionKind().Kind == "Role" && val.GetName() == obj.GetName() {
```

Comment:
Logically yes the Kind is ensured before this function is called. But since in the function containsRole obj is of type runtimeclient.Object, so to keep the function contract intact in the standalone call to the function i was checking the kind as well.

Reply:
Well, ok. but you don't have to check it inside of the for loop, but you could do it before that for loop, right? I mean at line 46

Reply:
sure, makes sense, I can do that.

Reply:
done